cts/cts-regression.in

```diff
@@ -1,230 +1,335 @@
+            values = [values]
+
+        components = set(values)
+
+        # If 'all', is found, replace it with the components it represents.
+        try:
+            components.remove('all')
+            components.update(set(all_components))
+        except KeyError:
+            pass
+
+        # Same for 'default'
+        try:
+            components.remove('default')
+            components.update(set(default_components))
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(components)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print.
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_components(components, verbose=False, valgrind=False):
+    """Run components' regression tests and report results for each.
+
+    :param components: A list of names of components for which to run
+        tests.
+    :type tests: list[str]
+    :return: :attr:`CrmExit.OK` if all tests were successful,
+        :attr:`CrmExit.ERROR` otherwise.
+    :rtype: :class:`CrmExit`
+    """
+    failed = []
+
+    for comp in components:
+        rc = comp.run(verbose, valgrind)
+        if rc != CrmExit.OK:
+            error_print('%s regression tests failed (%s)' % (comp.name, rc))
+            failed.append(comp.name)
+
+    if failed:
+        print('Failed regression tests:', end='')
+        for comp in failed:
+            print(' %s' % comp, end='')
+        print()
+        return CrmExit.ERROR
+
+    return CrmExit.OK
+
+
+def main():
+    """Run Pacemaker regression tests as specified by arguments."""
+    global TEST_HOME
+    try:
+        TEST_HOME = os.path.dirname(os.readlink(sys.argv[0]))
+    except OSError:
+        TEST_HOME = os.path.dirname(sys.argv[0])
+
+    # Available components
+    components = {
+        'cli': Component(
+            'cli',
+            'Command-line tools',
+            use_sudo=False,
+            supports_valgrind=True,
+        ),
+        'exec': Component(
+            'exec',
+            'Local resource agent executor',
+            use_sudo=True,
+            supports_valgrind=False,
+        ),
+        'fencing': Component(
+            'fencing',
+            'Fencer',
+            use_sudo=True,
+            supports_valgrind=False,
+        ),
+        'scheduler': Component(
+            'scheduler',
+            'Action scheduler',
+            use_sudo=False,
+            supports_valgrind=True,
+        ),
+    }
+
+    if REMOTE_ENABLED:
+        components['pacemaker_remote'] = Component(
+            'pacemaker_remote',
+            'Resource agent executor in remote mode',
+            use_sudo=True,
+            supports_valgrind=False,
+        )
+
+    # Build up program description
+    description = textwrap.dedent('''\
+        Run Pacemaker regression tests.
+
+        Available components (default components are 'scheduler cli'):
```

Comment:
note to self: might as well flip order to 'cli scheduler' now, since program description output is sorted